src/components/include/crt_blkpt.h

```diff
@@ -0,0 +1,298 @@
+#ifndef CRT_BLKPT_H
+#define CRT_BLKPT_H
+
+#include <cos_types.h>
+#include <ps.h>
+#include <sl.h>
+
+/***
+ * The event count/block point is an abstraction to synchronize the
+ * blocking behavior of different threads on abstract events. The
+ * events are usually tied to a specific state of another
+ * data-structure (into which the blkpt is embedded).  For example, a
+ * lock is taken and released thus generating an event for any
+ * blocking threads, or a ring buffer has a data item inserted into
+ * it, thus generating an event for any threads waiting for
+ * data. Concretely, we want a number of threads to be able to block,
+ * and a thread to be able to wake up one, or all of them. The
+ * challenge is solving a single race-condition:
+ *
+ * thd 0: check data-structure, determine the need for blocking and
+ *        waiting for an event
+ * thd 0: preemption, switching to thd 1
+ * thd 1: check data-structure, determine that an event is generated
+ * thd 1: call the scheduler, and wake all blocked threads (not
+ *        including thd 0 yet)
+ * thd 1: preempt, and switch to thd 0
+ * thd 0: call scheduler to block
+ *
+ * The resulting state is that thd 1 should have unblocked thd 0, but
+ * due to a race, the thd 0 will be blocked awaiting the *next* event
+ * that may never come. Event counts are meant to solve this
+ * problem. Traditional systems solve this problem using condition
+ * variables and a lock around the scheduling logic, but if you want
+ * to decouple the data-structure from the scheduler (e.g. as they are
+ * in different modes, or components), this is a fundamental problem.
+ *
+ * The event count abstraction:
+ *
+ * Assume the data-structure generating events has at least three
+ * states:
+ * S0: available
+ * S1: unavailable
+ * S2: unavailable & subscribed
+ *
+ * The transitions within the data-structure are:
+ * {S0->S1, S1->S0, S1->S2, S2->S0}
```

Comment:
I think there should be S2->S1 transitions, right? Suppose two threads are blocked on it, one thread unblocked, and there will still be one blocked on it so it is still unavailable... Or I have a misunderstanding here,